core/shared/src/main/scala/scalaz/zio/ExitResult.scala

```diff
@@ -10,32 +10,36 @@ sealed trait ExitResult[E, A] { self =>
   import ExitResult._
 
   final def succeeded: Boolean = self match {
-    case Completed(_)  => true
-    case Failed(_)     => false
-    case Terminated(_) => false
+    case Completed(_) => true
+    case _            => false
   }
 
   final def map[B](f: A => B): ExitResult[E, B] = self match {
-    case Completed(a)  => Completed(f(a))
-    case Failed(e)     => Failed(e)
-    case Terminated(t) => Terminated(t)
+    case Completed(a) => Completed(f(a))
+    case x            => x.asInstanceOf[ExitResult[E, B]]
   }
 
   final def mapError[E2](f: E => ExitResult[E2, A]): ExitResult[E2, A] = self match {
-    case ExitResult.Failed(e) => f(e)
-    case x                    => x.asInstanceOf[ExitResult[E2, A]]
+    case ExitResult.Failed(e, _) => f(e)
+    case x                       => x.asInstanceOf[ExitResult[E2, A]]
   }
 
   final def failed: Boolean = !succeeded
 
-  final def fold[Z](completed: A => Z, failed: E => Z, interrupted: Throwable => Z): Z = self match {
-    case Completed(v)  => completed(v)
-    case Failed(e)     => failed(e)
-    case Terminated(e) => interrupted(e)
-  }
+  final def fold[Z](completed: A => Z, failed: (E, List[Throwable]) => Z, interrupted: List[Throwable] => Z): Z =
+    self match {
+      case Completed(v)  => completed(v)
+      case Failed(e, ts) => failed(e, ts)
+      case Terminated(e) => interrupted(e)
+    }
 }
 object ExitResult {
-  final case class Completed[E, A](value: A)          extends ExitResult[E, A]
-  final case class Failed[E, A](error: E)             extends ExitResult[E, A]
-  final case class Terminated[E, A](error: Throwable) extends ExitResult[E, A]
+  final case class Completed[E, A](value: A)                       extends ExitResult[E, A]
+  final case class Failed[E, A](error: E, causes: List[Throwable]) extends ExitResult[E, A]
```

Comment:
These are not so much `causes` as a combination of finalizer defects (exceptions thrown from finalizers) and maybe interruption causes. This conflation makes me want to clean up the model and state:

 1. If `ExitResult.Failed`, then the fiber failed because it did not handle an `E` error, and in the process of terminating the fiber, finalizers were ran which may have resulted in some runtime exceptions. No interruption causes are reported because even if the fiber was interrupted at the same time it failed, ultimately the cause of its failure is the unhandled `E` error.
 2. If `ExitResult.Terminated`, then the fiber failed either because of interruption, or because of a defect, and any information on the failure is reported in `causes`.

It would be possible to further separate (2), but I'm not sure it's worth it.

Reply:
I admit I like the current `Failed` better. It's opinionated: the developer is accountable for exceptions, otherwise the whole program breaks down; there's no negotiation. If a finalizer throws, it may be a small thing, or it can have catastrophic consequences—I think the library should make the worst assumption.

And what to make of that asymmetry: if an `A` is produced in the value channel but finalizers throw exceptions, the action yields a `Terminated`.